DTS: Intel ioapic/mvic and gpio, uart, rtc...

tbursztyka opened the pull request:
A PR that enables a bit more the use of DTS for Intel SoCs as well as the related drivers such as uart ns16550 and qmsi gpio/rtc/uart ones

Note that only the 18 last patches are new, all the first ones are from PR #6406

Commit:
dts/i2c: Fixing copyright headers

Some yaml files were missing there copyright header.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/i2c: Cleanup Kconfig

Move driver specific to dedicated file when relevant (i.e.: more than
1-2 options), use if/endif also.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/i2c: Indent sources the same way everywhere

Using tabs only.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts: Add yaml descriptor for the DW i2c driver

Shared irq is not managed there as it is a very SoC specific issue.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts/x86: Add the i2c node in quark_x1000

This will generate the necessary configuration for the i2c controller on
quark_x1000.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
boards/galileo: Switch to DTS for generating i2c settings

Thus dropping SoC based generated information.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/i2c: Rename generic configure option in DW drivers

And adapt SoC's information relevantly.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts: Add yaml descriptor for the QMSI i2c driver

Used in some Intel quark SoC.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts/x86: Add the i2c node in intel curie

This will generate the necessary configuration for the 2 i2c controllers
on quark_se_c1000.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86: Enable i2c through DTS for quark_se

And adapt relevant boards accordingly.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts/x86: Add the i2c node in quark_d2000

This will generate the necessary configuration for the i2c controller
on quark_d2000.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86: Enable i2c through DTS for quark_d2000

And adapt the relevant board accordingly.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts: Add yaml descriptor for the QMSI SS i2c driver

Used in Intel quark_se's ARC core.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts/x86: Add the i2c node in Quark SE SS C1000

This will generate the necessary configuration for the 2 i2c controllers
on quark_se_c1000_ss.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/i2c: Disable some QMSI SS option when HAS_DTS_I2C is enabled

These options will be generated through DTS.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86: Enable i2c through DTS for quark_se SS

And adapt relevant boards accordingly.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/i2c: Removing useless configuration attribute in DW driver

Probably a left over form legacy interrupt handling.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/i2c: Use standard bitrate settings for DW driver

Applying the change to relevant arch/boards, either in their Kconfig or
the dts specific files.

Taking the opportunity in dw driver to rename the variable the same way
as they are everywhere else in the code (s/dev/dw and s/port/dev) in
init function.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/i2c: Use standard bitrate settings for QMSI driver

Apply the change on relevant arch/boards.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/i2c: Use standard bitrate settings for QMSI SS driver

Apply the change on relevant arch/boards.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts/i2c: Add support for QMSI SS interrupt declaration in nodes

i2c qmsi ss driver requires 4 interrupts: error, rx, tx and stop.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/i2c: Make QMSI SS driver IRQs set via CONFIG_ options

Now that dts i2 qmsi ss nodes generate the right options, let's use
them. Apply the relevant fixup on the targeted SoC.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts: Adding priority cell to Intel's IRQ controllers descriptions

Zephyr exposes the ability to set IRQ priority on MVIC and IOAPIC, so
let's enable this on dts level.
To keep compatibility with Linux, let's add the priority at the end
after sense. So imported dtsi with interrupt-cells set to 2 will work as
usual.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts/x86: Enable generating the IRQ priority on all SoCs

Now that MVIC and IOAPIC descripton support it, let's use it.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts: Add yaml descriptor for the QMSI GPIO driver

Used in Intel Quark SE/D2000 SoCs.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts/x86: Fix GPIO nodes for intel_curie and quark_d2000

Base addresses were wrong, and intel_curie missed a second gpio node.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86: Use dts to set gpio priorities for quark_se and quark_d2000

Get the name and irq flags generated through dts as well.
Fix Kconfig for the gpio driver accordingly.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch: Use dts to set rtc priorities for Intel quark, x86 and arc

Get the name generated through dts as well.
Fix the rtc driver and relevant SoCs accordingly.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts: Add yaml descriptor for the QMSI RTC driver

Used in Intel Quark SE/D2000 SoCs.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts/x86: Add Copyright headers to x86 dtsi files

All were missing it.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts/bindings: Remove useless attribute in QMSI uart node descriptor

Clock gate is set on driver level through QMSI API.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/uart: Use dts to set uart priorities for QMSI driver

Fix the qmsi uart driver and relevant SoCs accordingly.
Also: using config for irq everwhere relevantly and not an hardcoded
value in the driver.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts/x86: Fix UART nodes for ia32, atom and quark_x1000

Base addresses were wrong.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/uart: Use dts to set uart priorities for ns16550 driver

Fix the ns16550 uart driver and relevant SoCs accordingly.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts/arc: Add the GPIO nodes to quark_se_c1000_ss

There are 4 of them.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/arc: Use dts to set gpio priorities for quark_se_c1000_ss

Fix the qmsi ss gpio driver accordingly.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts: Add yaml descriptor for the QMSI SS GPIO driver

Used in Intel Quark SE ARC core.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
@erwango I removed you as a reviewer, it automatically added you because of the patches from PR #6406 which is touching some Kconfig stuff around stm32.

Commit:
dts/x86: Update i2c nodes with interrupts for quark_se and quark_d2000

Let's generate interrupt information for i2c through dts.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch: Use dts to set i2c priorities for quark_se/quark_d2000

Fix the qmsi i2c driver and the relevant SoCs accordingly.
Also applying relevant changes on quark_se_c1000_ss as it can use i2c
qmsi driver as well along with qmsi ss i2c driver.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts/arc: Add rtc node to quark_se_c1000_ss

RTC settings will be generated through dts, and this SoC owns an rtc
node.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

galak reviewed (changes requested):
So I see _ioapic_irq_set(irq, vector, flags) - I don’t see any means for priority to get set or passed down to the ioapic.  So I don’t see why we would add priority to the dts for iopaic.

Also, we should split this so mvic is one patch, and all fixes to mvic dts’s are in that patch.  If we have something meaningful for ioapic than that should be its own patch, with changes to .dts etc.

galak commented on drivers/gpio/Kconfig.qmsi:
```diff
@@ -50,12 +50,12 @@ config GPIO_QMSI_0
 config GPIO_QMSI_0_NAME
 	string "Driver name"
-	depends on GPIO_QMSI_0
+	depends on GPIO_QMSI_0 && !HAS_DTS
```
Are there any cases for !HAS_DTS?  Any reason to keep these Kconfig symbols?

tbursztyka replied:
Has it's been done in all other Kconfig places, config options there are kept with that dependency. 
Let's remove all these options when we are done porting everything to dts, and sure it works properly.

galak replied:
> Has it's been done in all other Kconfig places, config options there are kept with that dependency.
> Let's remove all these options when we are done porting everything to dts, and sure it works properly.

Its been the case that we've been removing Kconfig options if there aren't any uses for non-DTS.  So if its the case that its all users of GPIO_QMSI_0_NAME have HAS_DTS=y, than we should remove the Kconfig symbol.

tbursztyka commented:
@galak as @andrewboie mentioned here https://github.com/zephyrproject-rtos/zephyr/pull/6406#discussion_r172347803

Check include/arch/x86/arch.h

galak commented:
> @galak as @andrewboie mentioned here #6406 (comment)
> 
> Check include/arch/x86/arch.h

Based on include/arch/x86/arch.h - it seems as if mvic doesn't support priority, trying to understand the comments about how this work for ioapic.

tbursztyka commented:
Even though mvic seems to have hard-coded interrupt priorities, tightly related to the interrupt vectors, the priority is still inserted into the IDT, same core interrupt handling, same gen_idt tool used.

Does that mean mvic is handled the same way as ioapic @andrewboie ? Or is the priority in the IDT silently ignored in mvic case?

Commit:
drivers/gpio: Removing dts generated options in QMSI Kconfig

Now that all arch using QMSI gpio driver are generating the right
settings through DTS, these options can be removed from Kconfig.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

galak commented:
@andrewboie any comment on this:

> Even though mvic seems to have hard-coded interrupt priorities, tightly related to the interrupt vectors, the priority is still inserted into the IDT, same core interrupt handling, same gen_idt tool used.
> 
> Does that mean mvic is handled the same way as ioapic @andrewboie ? Or is the priority in the IDT silently ignored in mvic case?

tbursztyka commented:
@andrewboie we really need your input here :)

nashif commented:
interrupt vectors in the mvic are fixed and AFAIK you can't program them.

tbursztyka commented:
@nashif Sure but to keep things consistent within all i2c drivers and dts files, I needed no do these changes (so instead of a hard-coded value in driver, it's provided by dts from soc's dtsi). Now the issue is just about mvic priority.

tyrantlizard77 reviewed (commented):
On MVIC, it is basically a reduced feature set IOAPIC. IDT Vector assignments are fixed, they are just the IRQ line number + 32. So any priority value provided is just ignored. The only flexibility the OS has wrt priorities is to select a free vector for the timer interrupt (by setting its IRQ line in a register), and hence the priority of the timer. Everything else is graven in stone.

galak commented:
> On MVIC, it is basically a reduced feature set IOAPIC. IDT Vector assignments are fixed, they are just the IRQ line number + 32. So any priority value provided is just ignored. The only flexibility the OS has wrt priorities is to select a free vector for the timer interrupt (by setting its IRQ line in a register), and hence the priority of the timer. Everything else is graven in stone.

So all of that means that MVIC doesn't support priority and thus shouldn't be in the device tree for MVIC or MVIC related interrupts.
